### PR TITLE
SCP-5144 adapt dsl for runtime testing

### PR DESCRIPTION
**Pull request by paluh**
Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [ ] [Test report is updated](https://github.com/input-output-hk/marlowe-cardano/blob/main/marlowe/test/test-report.md) (if relevant)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
        - Review required
        - [x] Substantial changes to code, test, or documentation
        - [ ] Change made to Marlowe validator (@bwbush and @palas must be included as reviewers)
        - Review not required
        - [ ] Minor changes to non-critical code, documentation, nix derivations, configuration files, or scripts
        - [ ] Formatting, spelling, grammar, or reorganization
    - [x] Reviewer requested

### CODE REVIEW

**Comment by paluh** [comment]
Brian, I would like to add three points from my side:

* Unfortunately this PR also includes some changes to the `marlowe-apps`. I'm testing these changes when using the lib app in the runtime monitor component but I've not tested the command line tools themselves.
* This PR is pretty large - sorry about that. I'm to do pair review or summary of the implementation.
* Test resource cleanup is not reliable and done manually (token burning, faucet refunding). If we have failure in the runtime monitor thread or if a particular test fails we don't perform this action. I want to do a proper version of the cleanup by introducing `ResourceT` or something like that.

File: marlowe-cli/src/Language/Marlowe/CLI/Test/CLI/Interpret.hs [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+            (Just th, _) -> case anyCLIMarloweThread (mt, txBody) mTxIn th of
+              Just th' -> pure $ Just th'
+              Nothing  -> throwError "[AutoRun] Extending of the Marlowe thread failed."
+
+      thread' <- foldM step ciThread plan
+      let
+        marloweContract' = marloweContract { ciThread = thread' }
+      modifying (contractsL . coerced)  $ Map.insert coContractNickname marloweContract'
+    SimulationMode -> do
+      -- TODO: We should be able to run balancing in here even in the simulation mode
```
Comment (bwbush) [inline marlowe-cli/src/Language/Marlowe/CLI/Test/CLI/Interpret.hs:440]: Yes, `Language.Marlowe.Runtime.Transaction.Safety` does this already.

Comment (paluh) [reply marlowe-cli/src/Language/Marlowe/CLI/Test/CLI/Interpret.hs:440]: @bwbush I created a ticket for this (PLT-5583). May I resolve this for now and proceed and merge?

**Review by bwbush — APPROVED** [review]
This is great, and so very useful! Thanks for refactoring the apps and cli modules, adding thread support, etc. for cli and runtime testing.

I think we should merge this now and, if needed, follow up with small PRs.

**Pull request merged**